### PR TITLE
CBL-6320: Test multi level array index - Low Level

### PR DESCRIPTION
**Pull request by jianminzhao**
Added following tests:
- Deeply nested arrays: 10 levels deep with and without index.
- Triggers of unnest tables (for array indexes)
  * insertion triggers
  * deletion triggers
  * update triggers
- Edge case, missing property at the unnestPath.

### CODE REVIEW

File: C/tests/c4QueryTest.cc [hunk 1]
```diff
@@ -985,6 +985,48 @@ N_WAY_TEST_CASE_METHOD(NestedQueryTest, "C4Query Nested UNNEST", "[Query][C]") {
...
+            REQUIRE(c4db_createIndex2(db, C4STR("contacts"), C4STR(""), kC4N1QLQuery, kC4ArrayIndex, &indexOpts,
+                                      nullptr));
+        }
+
+        const char* queryStr = "SELECT doc.name, contact FROM _default AS doc UNNEST doc.contacts AS contact";
+
+        // For the above query, server returns
+        //         vector<string> results {
+        // "Jonh Doe, Contact1",
+        // "Jonh Doe, Contact2"
```
Comment (pasin) [inline C/tests/c4QueryTest.cc:1014]: The indentation appears to be off here.

File: C/tests/c4QueryTest.cc [hunk 2]
```diff
@@ -985,6 +985,48 @@ N_WAY_TEST_CASE_METHOD(NestedQueryTest, "C4Query Nested UNNEST", "[Query][C]") {
...
+    deleteDatabase();
+    db = c4db_openNamed(kDatabaseName, &dbConfig(), ERROR_INFO());
+    std::stringstream documents{R"(
+[
+{"name": "Jonh Doe", "contacts": ["Contact1", "Contact2"]},
+{"name": "Scott Tiger", "contacts": []},
+{"name": "Foo Bar"},
+{"name": "Harry Potter", "contacts": "my contacts" }
+]
+)"};
```
Comment (pasin) [inline C/tests/c4QueryTest.cc:998]: The indentation appears to be incorrect here.

File: LiteCore/tests/QueryTest.cc [hunk 3]
```diff
@@ -145,6 +145,366 @@ N_WAY_TEST_CASE_METHOD(QueryTest, "Create/Delete Array Index (Multi-level)", "[Q
...
+        for ( int i = 0; i <= top; ++i ) {
+            if ( i == 0 && !docID.empty() ) ss << docID << "-";
+            if ( i > 0 ) ss << "-";
+            ss << letters[i].first << letters[i].second;
+        }
+        return ss.str();
+    }
+
+    string multiLevelArrayIter(std::vector<std::pair<char, unsigned>>& letters, int& letter, int arrDim,
+                               const string& docID) {
```
Comment (pasin) [inline LiteCore/tests/QueryTest.cc:178]: Can you provide comment with an example of the result that this function generates?

File: LiteCore/tests/QueryTest.cc [hunk 4]
```diff
@@ -145,6 +145,366 @@ N_WAY_TEST_CASE_METHOD(QueryTest, "Create/Delete Array Index (Multi-level)", "[Q
...
+            auto pos2 = updateJson.find("\"", pos + 1);
+            Assert(pos2 != string::npos);
+            ++pos2;
+            for ( size_t i = pos + 8; i < pos2; ++i ) {
+                if ( 'a' <= updateJson[i] && updateJson[i] <= 'z' ) { updateJson[i] = 'A' + updateJson[i] - 'a'; }
+            }
+            arrayElements.push_back(updateJson.substr(pos + 1, pos2 - pos - 2));
+            pos = pos2;
+        }
+    }
```
Comment (pasin) [inline LiteCore/tests/QueryTest.cc:481]: Instead of this loop to update string, maybe modify the json generator to specify the format (e.g. lowercase or uppercase)?

File: C/tests/c4QueryTest.cc (hunk 1 above)
Comment (jianminzhao) [reply C/tests/c4QueryTest.cc:1014]: will fix

File: C/tests/c4QueryTest.cc (hunk 2 above)
Comment (jianminzhao) [reply C/tests/c4QueryTest.cc:998]: will fix

File: C/tests/c4QueryTest.cc [hunk 5]
```diff
@@ -911,7 +911,7 @@ N_WAY_TEST_CASE_METHOD(NestedQueryTest, "C4Query Nested UNNEST", "[Query][C]") {
             "Univ of Pennsylvania, student_112, 3, piano",  "Univ of Pennsylvania, student_112, 3, swimming",
             "Univ of Pennsylvania, student_189, 2, violin", "Univ of Pennsylvania, student_189, 2, movies"};
 
-    for ( int withIndex = 1; withIndex <= 1; ++withIndex ) {
+    for ( int withIndex = 0; withIndex <= 1; ++withIndex ) {
```
Comment (borrrden) [inline C/tests/c4QueryTest.cc:914]: Consider using Catch `GENERATE` macro

**Review by borrrden — COMMENTED** [review]
Not sure I am familiar enough with this implementation to do a proper review

File: LiteCore/tests/QueryTest.cc [hunk 6]
```diff
@@ -145,6 +145,366 @@ N_WAY_TEST_CASE_METHOD(QueryTest, "Create/Delete Array Index (Multi-level)", "[Q
...
+            AddDocWithJSON(store, docID, t, json.c_str());
+        }
+        t.commit();
+        return ret;
+    }
+
+    // Pre-conditions: 'a' <= a && letterDim <= 'a' +26 - a
+    // Example output:  "doc001-a0-b0", "doc001-a0-b1", "doc001-a1-b0", "doc001-a1-b1", "doc002-a0-b0"
+    // , "doc002-a0-b1", "doc002-a1-b0", "doc002-a1-b1", ...
+    vector<string> unnestedElements(const vector<string>& docs, char a, unsigned letterDim, unsigned arrDim) {
```
Comment (callumbirks) [inline LiteCore/tests/QueryTest.cc:248]: I think it could be nice to describe in the comment what the function parameters are, and what the function is actually doing (what does "unnestedElements" mean). I don't know what `letterDim`, `arrayDim`, or `a` are supposed to be.

**Review by callumbirks — CHANGES_REQUESTED** [review]
I think we could do with some more comments on functions. Other than that, looks fine - with the caveat that, like Jim, I don't understand enough about the implementation to give a thorough review.

File: LiteCore/tests/QueryTest.cc (hunk 4 above)
Comment (jianminzhao) [reply LiteCore/tests/QueryTest.cc:481]: little awkward. I want to change the case for the string literals, like "a0-b0-c0", but not the property names. So I just did it hard way. It's one time need.

File: C/tests/c4QueryTest.cc (hunk 5 above)
Comment (jianminzhao) [reply C/tests/c4QueryTest.cc:914]: This style is apparently dominant for the batch of tests of this nature. It loops twice, first time w/o index, the second time w/ the index. It's also natural because it applies to the same db, same query, and expecting same result.

File: LiteCore/tests/QueryTest.cc [hunk 7]
```diff
@@ -145,6 +145,366 @@ N_WAY_TEST_CASE_METHOD(QueryTest, "Create/Delete Array Index (Multi-level)", "[Q
...
+              s        = multiLevelArrayIter(letters, letter, arrDim, docID) ) {
+            ret = ret + s;
+        }
+        return ret;
+    }
+
+    // build a json of nested arrays, like:
+    // {"a":[{"b":["doc001-a0-b0","doc001-a0-b1"]},{"b":["doc001-a1-b0","doc001-a1-b1"]}]}
+    // Pre-conditions: 'a' <= a && letterDim <= 'a' +26 - a
+    string buildMultiLevelArray(char a, int letterDim, int arrDim, const string& docID) {
```
[the remark this answers is not in the document]
Comment (jianminzhao) [reply LiteCore/tests/QueryTest.cc:222]: will do

**Review by pasin — APPROVED** [review]

**Review by callumbirks — APPROVED** [review]
